junebug/logging_service.py

```diff
@@ -122,3 +124,72 @@ def stopService(self):
 
     def registered(self):
         return self.log_observer in self.logger.observers
+
+
+def reverse_read(filename, buf):
+    '''
+    Read the non-blank lines from a file in reverse order.
+
+    :param str filename: The path of the file to be read.
+    :param int buf: The size of the read buffer.
+
+    :returns:
+        A generator that yields each line from the file in reverse order as a
+        string.
+    '''
+    with open(filename) as f:
+        f.seek(0, os.SEEK_END)
+        total_size = remaining_size = f.tell()
+        offset = 0
+        incomplete_line = None
+        while remaining_size > 0:
+            offset = min(total_size, offset + buf)
+            f.seek(total_size - offset, os.SEEK_SET)
+            data = f.read(min(remaining_size, buf))
+            # If we read no data, we should exit the loop
+            if len(data) == 0:
```

Comment:
So, a couple more thoughts around this:
* I think we should ditch the `total_size` -- we only ever care about `remaining_size` (i.e. where we last read from). Removing `total_size` should make the logic a bit clearer.
* There is a corner case we're not covering which is that if `f.read()` returns fewer bytes than requested, we won't have all the data up to where we last read from. I seem to remember that the blocking Python `.read` call makes multiple calls to the underlying C `fread` until the number of bytes requested have been returned or the file ends, but I don't have the Python source checked out right now. In short, I think we are fine as we are, but just wanted to mention the issue.